packages/loopring_v3/contracts/impl/libexchange/ExchangeBlocks.sol

```diff
@@ -374,50 +263,238 @@ library ExchangeBlocks
+        ExchangeData.Block memory firstBlock;
+        uint[] memory batch = new uint[](blocks.length);
+        while (numBlocksVerified < blocks.length) {
+            // Find all blocks of the same type
+            uint batchLength = 0;
+            for (uint i = 0; i < blocks.length; i++) {
+                if (blockVerified[i] == false) {
+                    if (batchLength == 0) {
+                        firstBlock = blocks[i];
+                        batch[batchLength++] = i;
+                    } else {
+                        ExchangeData.Block memory _block = blocks[i];
+                        if (_block.blockType == firstBlock.blockType &&
+                            _block.blockSize == firstBlock.blockSize &&
+                            _block.blockVersion == firstBlock.blockVersion) {
+                            batch[batchLength++] = i;
+                        }
+                    }
+                }
+            }
+
+            // Prepare the data for batch verification
+            uint[] memory publicInputs = new uint[](batchLength);
+            uint[] memory proofs = new uint[](batchLength * 8);
+            for (uint i = 0; i < batchLength; i++) {
+                uint blockIdx = batch[i];
+                // Mark the block as verified
+                blockVerified[blockIdx] = true;
+                // Strip the 3 least significant bits of the public data hash
+                // so we don't have any overflow in the snark field
+                publicInputs[i] = uint(publicDataHashes[blockIdx]) >> 3;
+                // Copy proof
+                ExchangeData.Block memory _block = blocks[blockIdx];
+                for (uint j = 0; j < 8; j++) {
+                    proofs[i*8 + j] = _block.proof[j];
+                }
+            }
+
+            // Verify the proofs
+            require(
+                S.blockVerifier.verifyProofs(
+                    uint8(firstBlock.blockType),
+                    S.onchainDataAvailability,
+                    firstBlock.blockSize,
+                    firstBlock.blockVersion,
+                    publicInputs,
+                    proofs
+                ),
+                "INVALID_PROOF"
+            );
+
+            numBlocksVerified += batchLength;
+        }
+    }
+
+    function validateConditionalTransfers(
+        ExchangeData.State storage S,
+        bytes  memory data,
+        bytes  memory auxiliaryData
+        )
+        private
+    {
+        uint offset = 4 + 32 + 32 + 32;
+        // The length of the auxiliary data needs to match the number of conditional transfers
+        uint numConditionalTransfers = data.bytesToUint32(offset);
+        require(auxiliaryData.length == numConditionalTransfers * 4, "INVALID_AUXILIARYDATA_LENGTH");
+        offset += 4;
+
+        uint24 operatorAccountID = data.bytesToUint24(offset);
+        offset += 3;
+
+        if (numConditionalTransfers > 0) {
```

Comment:
move this if-statement right after `uint numConditionalTransfers = data.bytesToUint32(offset);`